contracts/sol6/fprConversionRates/SimpleVolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,116 @@
+pragma solidity 0.6.6;
+
+import "../IERC20.sol";
+import "../utils/Withdrawable3.sol";
+
+contract SimpleVolumeImbalanceRecorder is Withdrawable3 {
+    uint256 internal constant SLIDING_WINDOW_SIZE = 5;
+    uint256 internal constant POW_2_64 = 2**64;
+
+    struct TokenControlInfo {
+        uint128 minimalRecordResolution; // can be roughly 1 cent
+        uint64 maxPerBlockImbalanceInResolution; // in resolution
+        uint64 maxTotalImbalanceInResolution; // in resolution
+    }
+
+    /// @dev data can compress into 1 word in storage
+    struct TokenImbalanceData {
+        int64 lastBlockBuyUnitsImbalance;
+        uint64 lastBlock;
+        int64 totalBuyUnitsImbalance;
+        uint64 lastRateUpdateBlock;
+    }
+
+    mapping(IERC20 => TokenControlInfo) internal tokenControlInfos;
+    mapping(IERC20 => TokenImbalanceData) public tokenImbalanceDatas;
+
+    constructor(address _admin) public Withdrawable3(_admin) {}
+
+    function setTokenControlInfo(
+        IERC20 token,
+        uint128 minimalRecordResolution,
+        uint256 maxPerBlockImbalance,
+        uint256 maxTotalImbalance
+    ) public onlyAdmin {
+        uint256 maxPerBlockImbalanceInResolution = maxPerBlockImbalance /
+            uint256(minimalRecordResolution);
+        require(maxPerBlockImbalanceInResolution <= 2**63 - 1, "overflow maxPerBlockImbalance");
+        uint256 maxTotalImbalanceInResolution = maxTotalImbalance /
+            uint256(minimalRecordResolution);
+        require(maxTotalImbalanceInResolution <= 2**63 - 1, "overflow maxTotalImbalance");
+        tokenControlInfos[token] = TokenControlInfo(
+            uint128(minimalRecordResolution),
+            uint64(maxPerBlockImbalanceInResolution),
+            uint64(maxTotalImbalanceInResolution)
+        );
+    }
+
+    function getTokenControlInfo(IERC20 token)
+        public
+        view
+        returns (
+            uint256 minimalRecordResolution,
+            uint256 maxPerBlockImbalanceInResolution,
+            uint256 maxTotalImbalanceInResolution
+        )
+    {
+        minimalRecordResolution = tokenControlInfos[token].minimalRecordResolution;
+        maxPerBlockImbalanceInResolution = tokenControlInfos[token]
+            .maxPerBlockImbalanceInResolution;
+        maxTotalImbalanceInResolution = tokenControlInfos[token].maxTotalImbalanceInResolution;
+    }
+
+    function addImbalance(
+        IERC20 token,
+        int256 buyAmount,
+        uint256 rateUpdateBlock,
+        uint256 currentBlock
+    ) internal {
+        int64 recordedBuyAmount = int64(
+            buyAmount / int256(tokenControlInfos[token].minimalRecordResolution)
+        );
+        TokenImbalanceData memory currentBlockData = tokenImbalanceDatas[token];
+
+        // first scenario - this is not the first tx in the current block
+        if (currentBlockData.lastBlock == currentBlock) {
+            currentBlockData.lastBlockBuyUnitsImbalance += recordedBuyAmount;
```

Comment:
seems we should add overflow checks in this function.
not sure why we didn't have it.